static/app/stores/persistedStore.tsx

```diff
@@ -0,0 +1,121 @@
+import React, {
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from 'react';
+
+import useApi from 'sentry/utils/useApi';
+import useOrganization from 'sentry/utils/useOrganization';
+import {OnboardingState} from 'sentry/views/onboarding/targetedOnboarding/types';
+
+import OrganizationStore from './organizationStore';
+import {useLegacyStore} from './useLegacyStore';
+
+type PersistedStore = Readonly<{
+  onboarding: Partial<OnboardingState> | null;
+}>;
+
+const DefaultPersistedStore: PersistedStore = {
+  onboarding: null,
+};
+
+type PersistedStoreContextValue = [
+  PersistedStore | null,
+  React.Dispatch<React.SetStateAction<PersistedStore | null>>
+];
+
+export const PersistedStoreContext = createContext<PersistedStoreContextValue | null>(
+  null
+);
+
+function usePersistedStore(): PersistedStoreContextValue {
+  const context = useContext(PersistedStoreContext);
+
+  if (!context) {
+    throw new Error('usePersistedStore was called outside of PersistedStoreProvider');
+  }
+
+  return context;
+}
+
+// Client-only state with TTL persisted on the server side in a redis store.
+export function PersistedStoreProvider(props: {children: React.ReactNode}) {
+  const [state, setState] = useState<PersistedStore | null>(null);
+
+  const api = useApi();
+  const {organization} = useLegacyStore(OrganizationStore);
+
+  useEffect(() => {
+    if (!organization) {
```

Comment:
@Neo-Zhixing should we log a warning or something?

Reply:
I don't think we should. It's entirely possible that the `PersistedStoreProvider` gets loaded before `OrganizationState` does. When `OrganizationState` loads the `useEffect` hook will rerun since we've pass `organization` in the dependency list.